Unicode error in `astropy.utils.data` in Python 2.7

Hi, I'm not sure it is a `astropy` bug, I am implementing a feature in `healpy` using `astropy.utils.data`, it works great in Python 3, but it fails in Python 2.7 on Travis, while it works fine on my machine.

It is a unicode error in `download_file`:

```healpy/test/test_pixelweights.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
healpy/sphtfunc.py:167: in map2alm
    package="healpy")
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/astropy/utils/data.py:636: in get_pkg_data_filename
    timeout=remote_timeout)
../../../virtualenv/python2.7.14/lib/python2.7/site-packages/astropy/utils/data.py:1074: in download_file
    if url_mirror in url2hash:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <[ValueError("invalid operation on closed shelf") raised in repr()] DbfilenameShelf object at 0x7f28d539cbd8>
key = 'http://www.astropy.org/astropy-data/full_weights/healpix_full_weights_nside_0064.fits'
    def __contains__(self, key):
>       return key in self.dict
E       TypeError: gdbm key must be string, not unicode
```

see the full travis log here: https://travis-ci.org/healpy/healpy/jobs/382367728

here is the pull request implementing this feature: https://github.com/healpy/healpy/pull/442

would you have a suggestion on how to debug this? thanks!!

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -81,6 +81,9 @@ astropy.units
 astropy.utils
 ^^^^^^^^^^^^^
 
+- Fixed a bug that improperly handled unicode case of URL mirror in Python 2.
+  [#7493]
+
 astropy.visualization
 ^^^^^^^^^^^^^^^^^^^^^
 

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -97,6 +97,7 @@ astropy.wcs
 Other Changes and Additions
 ---------------------------
 
+- Bundled ``pytest-remotedata`` plugin is upgraded to 0.3. [#7493]
 
 
 2.0.6 (2018-04-23)

--- a/astropy/extern/plugins/pytest_remotedata/disable_internet.py
+++ b/astropy/extern/plugins/pytest_remotedata/disable_internet.py
@@ -12,7 +12,9 @@ socket_create_connection = socket.create_connection
 socket_bind = socket.socket.bind
 socket_connect = socket.socket.connect
 
-ASTROPY_HOSTS = ['data.astropy.org', 'astropy.stsci.edu', 'www.astropy.org']
+GITHUB_HOSTS = ['www.github.io']
+ASTROPY_HOSTS = (['data.astropy.org', 'astropy.stsci.edu', 'www.astropy.org'] +
+                 GITHUB_HOSTS)
 
 INTERNET_OFF = False
 
@@ -22,14 +24,31 @@ INTERNET_OFF = False
 _orig_opener = None
 
 
+def _resolve_host_ips(hostname, port=80):
+    """
+    Obtain all the IPs, including aliases, in a way that supports
+    IPv4/v6 dual stack.
+    """
+    try:
+        ips = set([s[-1][0] for s in socket.getaddrinfo(hostname, port)])
+    except socket.gaierror:
+        ips = set([])
+
+    ips.add(hostname)
+    return ips
+
+
 # ::1 is apparently another valid name for localhost?
 # it is returned by getaddrinfo when that function is given localhost
 
-def check_internet_off(original_function, allow_astropy_data=False):
+def check_internet_off(original_function, allow_astropy_data=False,
+                       allow_github_data=False):
     """
     Wraps ``original_function``, which in most cases is assumed
     to be a `socket.socket` method, to raise an `IOError` for any operations
     on non-local AF_INET sockets.
+
+    Allowing Astropy data also automatically allow GitHub data.
     """
 
     def new_function(*args, **kwargs):
@@ -41,7 +60,7 @@ def check_internet_off(original_function, allow_astropy_data=False):
                 return original_function(*args, **kwargs)
             host = args[1][0]
             addr_arg = 1
-            valid_hosts = ('localhost', '127.0.0.1', '::1')
+            valid_hosts = set(['localhost', '127.0.0.1', '::1'])
         else:
             # The only other function this is used to wrap currently is
             # socket.create_connection, which should be passed a 2-tuple, but
@@ -51,22 +70,30 @@ def check_internet_off(original_function, allow_astropy_data=False):
 
             host = args[0][0]
             addr_arg = 0
-            valid_hosts = ('localhost', '127.0.0.1')
+            valid_hosts = set(['localhost', '127.0.0.1'])
 
+        # Astropy + GitHub data
         if allow_astropy_data:
             for valid_host in ASTROPY_HOSTS:
-                valid_host_ip = socket.gethostbyname(valid_host)
-                valid_hosts += (valid_host, valid_host_ip)
+                valid_hosts = valid_hosts.union(_resolve_host_ips(valid_host))
+
+        # Only GitHub data
+        elif allow_github_data:
+            for valid_host in GITHUB_HOSTS:
+                valid_hosts = valid_hosts.union(_resolve_host_ips(valid_host))
 
         hostname = socket.gethostname()
         fqdn = socket.getfqdn()
 
         if host in (hostname, fqdn):
             host = 'localhost'
+            host_ips = set([host])
             new_addr = (host, args[addr_arg][1])
             args = args[:addr_arg] + (new_addr,) + args[addr_arg + 1:]
+        else:
+            host_ips = _resolve_host_ips(host)
 
-        if any(h in host for h in valid_hosts):
+        if len(host_ips & valid_hosts) > 0:  # Any overlap is acceptable
             return original_function(*args, **kwargs)
         else:
             raise IOError("An attempt was made to connect to the internet "
@@ -75,7 +102,8 @@ def check_internet_off(original_function, allow_astropy_data=False):
     return new_function
 
 
-def turn_off_internet(verbose=False, allow_astropy_data=False):
+def turn_off_internet(verbose=False, allow_astropy_data=False,
+                      allow_github_data=False):
     """
     Disable internet access via python by preventing connections from being
     created using the socket module.  Presumably this could be worked around by
@@ -103,9 +131,15 @@ def turn_off_internet(verbose=False, allow_astropy_data=False):
     opener = urllib.request.build_opener(no_proxy_handler)
     urllib.request.install_opener(opener)
 
-    socket.create_connection = check_internet_off(socket_create_connection, allow_astropy_data=allow_astropy_data)
-    socket.socket.bind = check_internet_off(socket_bind, allow_astropy_data=allow_astropy_data)
-    socket.socket.connect = check_internet_off(socket_connect, allow_astropy_data=allow_astropy_data)
+    socket.create_connection = check_internet_off(
+        socket_create_connection, allow_astropy_data=allow_astropy_data,
+        allow_github_data=allow_github_data)
+    socket.socket.bind = check_internet_off(
+        socket_bind, allow_astropy_data=allow_astropy_data,
+        allow_github_data=allow_github_data)
+    socket.socket.connect = check_internet_off(
+        socket_connect, allow_astropy_data=allow_astropy_data,
+        allow_github_data=allow_github_data)
 
     return socket
 

--- a/astropy/extern/plugins/pytest_remotedata/plugin.py
+++ b/astropy/extern/plugins/pytest_remotedata/plugin.py
@@ -16,32 +16,37 @@ def pytest_addoption(parser):
         "--remote-data", nargs="?", const='any', default='none',
         help="run tests with online data")
 
-    parser.addini('remote_data_strict',
+    parser.addini(
+        'remote_data_strict',
         "If 'True', tests will fail if they attempt to access the internet "
         "but are not explicitly marked with 'remote_data'",
         type="bool", default=False)
-
 
 
 def pytest_configure(config):
     config.getini('markers').append(
         'remote_data: Apply to tests that require data from remote servers')
     config.getini('markers').append(
-        'internet_off: Apply to tests that should only run when network access is deactivated')
+        'internet_off: Apply to tests that should only run when '
+        'network access is deactivated')
 
     strict_check = bool(config.getini('remote_data_strict'))
 
     remote_data = config.getoption('remote_data')
-    if remote_data not in ['astropy', 'any', 'none']:
+    if remote_data not in ['astropy', 'any', 'github', 'none']:
         raise pytest.UsageError(
             "'{}' is not a valid source for remote data".format(remote_data))
 
     # Monkeypatch to deny access to remote resources unless explicitly told
     # otherwise
-    if strict_check and remote_data != 'any':
-        turn_off_internet(
-            verbose=config.option.verbose,
-            allow_astropy_data=(remote_data == 'astropy'))
+    if strict_check:
+        if remote_data == 'github':  # GitHub only
+            turn_off_internet(
+                verbose=config.option.verbose, allow_github_data=True)
+        elif remote_data != 'any':  # Astropy + GitHub or nothing
+            turn_off_internet(
+                verbose=config.option.verbose,
+                allow_astropy_data=(remote_data == 'astropy'))
 
 
 def pytest_unconfigure():
@@ -65,13 +70,23 @@ def pytest_runtest_setup(item):
         raise ValueError("remote_data and internet_off are not compatible")
 
     if remote_data is not None:
-        source = remote_data.kwargs.get('source', 'any')
-        if source not in ('astropy', 'any'):
-            raise ValueError("source should be 'astropy' or 'any'")
+        if len(remote_data.args) > 0:
+            source = remote_data.args[0]
+        else:
+            source = remote_data.kwargs.get('source', 'any')
+
+        if source not in ('astropy', 'any', 'github'):
+            raise ValueError("source should be 'astropy', 'any', or 'github'")
 
         if remote_data_config == 'none':
             pytest.skip("need --remote-data option to run")
+        elif remote_data_config == 'github':
+            if source in ('any', 'astropy'):
+                pytest.skip("need --remote-data or "
+                            "--remote-data=astropy option to run")
         elif remote_data_config == 'astropy':
+            # When --remote-data=astropy is given, skip tests simply
+            # marked as --remote-data
             if source == 'any':
                 pytest.skip("need --remote-data option to run")
 

--- a/astropy/utils/data.py
+++ b/astropy/utils/data.py
@@ -1071,6 +1071,9 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None):
                         if url_key.startswith(cur_url):
                             url_mirror = url_key.replace(cur_url,
                                                          conf.dataurl_mirror)
+                            if six.PY2 and isinstance(url_mirror,
+                                                      six.text_type):
+                                url_mirror = url_mirror.encode('utf-8')
                             if url_mirror in url2hash:
                                 return url2hash[url_mirror]
 

--- a/astropy/utils/tests/test_data.py
+++ b/astropy/utils/tests/test_data.py
@@ -50,7 +50,7 @@ else:
     HAS_PATHLIB = True
 
 
-@remote_data('astropy')
+@remote_data(source='astropy')
 def test_download_nocache():
     from ..data import download_file
 
@@ -58,7 +58,7 @@ def test_download_nocache():
     assert os.path.isfile(fnout)
 
 
-@remote_data('astropy')
+@remote_data(source='astropy')
 def test_download_parallel():
     from ..data import (download_file, download_files_in_parallel,
                         _get_download_cache_locs, get_cached_urls,
@@ -84,7 +84,7 @@ def test_download_parallel():
             (mirror_url in c_urls) and (main_url not in c_urls))
 
 
-@remote_data('astropy')
+@remote_data(source='astropy')
 def test_download_noprogress():
     from ..data import download_file
 
@@ -92,7 +92,7 @@ def test_download_noprogress():
     assert os.path.isfile(fnout)
 
 
-@remote_data('astropy')
+@remote_data(source='astropy')
 def test_download_cache():
 
     from ..data import download_file, clear_download_cache
@@ -125,7 +125,7 @@ def test_download_cache():
     assert not os.path.isdir(lockdir), 'Cache dir lock was not released!'
 
 
-@remote_data('astropy')
+@remote_data(source='astropy')
 def test_url_nocache():
 
     from ..data import get_readable_fileobj
@@ -134,7 +134,7 @@ def test_url_nocache():
         assert page.read().find('Astropy') > -1
 
 
-@remote_data('astropy')
+@remote_data(source='astropy')
 def test_find_by_hash():
 
     from ..data import get_readable_fileobj, get_pkg_data_filename, clear_download_cache
@@ -153,7 +153,7 @@ def test_find_by_hash():
     assert not os.path.isdir(lockdir), 'Cache dir lock was not released!'
 
 
-@remote_data('astropy')
+@remote_data(source='astropy')
 def test_find_invalid():
     from ..data import get_pkg_data_filename
 
@@ -299,7 +299,7 @@ def test_get_pkg_data_contents():
     assert contents1 == contents2
 
 
-@remote_data('astropy')
+@remote_data(source='astropy')
 def test_data_noastropy_fallback(monkeypatch):
     """
     Tests to make sure the default behavior when the cache directory can't
@@ -436,7 +436,7 @@ def test_compressed_stream():
         assert f.read().rstrip() == b'CONTENT'
 
 
-@remote_data('astropy')
+@remote_data(source='astropy')
 def test_invalid_location_download():
     """
     checks that download_file gives a URLError and not an AttributeError,
@@ -459,7 +459,7 @@ def test_invalid_location_download_noconnect():
         download_file('http://astropy.org/nonexistentfile')
 
 
-@remote_data('astropy')
+@remote_data(source='astropy')
 def test_is_url_in_cache():
     from ..data import download_file, is_url_in_cache
 

